crates/wdl-engine/src/backend/slurm_apptainer.rs

```diff
@@ -0,0 +1,886 @@
+use tokio::io::BufReader;
+use tokio::process::Command;
+use tokio::sync::broadcast;
+use tokio_util::sync::CancellationToken;
+use tracing::error;
+use tracing::trace;
+use tracing::warn;
+
+use super::ApptainerConfig;
+use super::COMMAND_FILE_NAME;
+use super::TaskExecutionBackend;
+use super::TaskManager;
+use super::TaskManagerRequest;
+use super::TaskSpawnRequest;
+use super::WORK_DIR_NAME;
+use crate::ONE_GIBIBYTE;
+use crate::PrimitiveValue;
+use crate::STDERR_FILE_NAME;
+use crate::STDOUT_FILE_NAME;
+use crate::TaskExecutionResult;
+use crate::Value;
+use crate::config::Config;
+use crate::config::TaskResourceLimitBehavior;
+use crate::path::EvaluationPath;
+use crate::v1;
+
+/// The name of the file where the Apptainer command invocation will be written.
+const APPTAINER_COMMAND_FILE_NAME: &str = "apptainer_command";
+
+/// The root guest path for inputs.
+const GUEST_INPUTS_DIR: &str = "/mnt/task/inputs/";
+
+/// The maximum length of a Slurm job name.
+// TODO ACF 2025-10-13: I worked this out experimentally on the cluster I happen
+// to have access to. I do not know whether this translates to other Slurm
+// installations, and cannot find documentation about what this limit should be
+// or whether it's configurable.
+const SLURM_JOB_NAME_MAX_LENGTH: usize = 1024;
+
+/// A request to execute a task on a Slurm + Apptainer backend.
+#[derive(Debug)]
+struct SlurmApptainerTaskRequest {
+    /// The desired configuration of the backend.
+    backend_config: Arc<SlurmApptainerBackendConfig>,
+    /// The name of the task, potentially truncated to fit within the Slurm job
+    /// name length limit.
+    name: String,
+    /// The task spawn request.
+    spawn_request: TaskSpawnRequest,
+    /// The requested container for the task.
+    container: String,
+    /// The requested CPU reservation for the task.
+    required_cpu: f64,
+    /// The requested memory reservation for the task.
+    required_memory: ByteSize,
+    /// The broadcast channel to update interested parties with the status of
+    /// executing tasks.
+    ///
+    /// This backend does not yet take advantage of the full Crankshaft
+    /// machinery, but we send rudimentary messages on this channel which helps
+    /// with UI presentation.
+    crankshaft_events: Option<broadcast::Sender<Event>>,
+    /// The cancellation token for this task execution request.
+    cancellation_token: CancellationToken,
+}
+
+impl TaskManagerRequest for SlurmApptainerTaskRequest {
+    fn cpu(&self) -> f64 {
+        self.required_cpu
+    }
+
+    fn memory(&self) -> u64 {
+        self.required_memory.as_u64()
+    }
+
+    async fn run(self) -> anyhow::Result<super::TaskExecutionResult> {
+        let crankshaft_task_id = crankshaft::events::next_task_id();
+
+        let attempt_dir = self.spawn_request.attempt_dir();
+
+        // Create the host directory that will be mapped to the WDL working directory.
+        let wdl_work_dir = attempt_dir.join(WORK_DIR_NAME);
```

Comment:
nit: pulling out the process of creating a file, maybe setting it's permissions, and maybe writing contents to it would make the first part of this method more readable and shrink it down significantly.